### PR TITLE
extending 'tf.keras.Model.evaluate' API to permit using class_weight directly.

### PR DESCRIPTION
**Pull request by ayushmankumar7**
Issue : #35825

### CODE REVIEW

**Review by qlzh727 — CHANGES_REQUESTED** [review]
The class weights is not used at all in this PR, could u update the data_handler to use it?

**Comment by ayushmankumar7** [comment]
Oh yes. Sorry, i missed that earlier. 
@qlzh727  I have made the changes.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Could u update unit test to make sure the change take effect?

Also since this change touch the API surface, could u also update the API golden file?

**Comment by ayushmankumar7** [comment]
@qlzh727  #37463 I am facing this issue while performing the unit test

**Comment by ayushmankumar7** [comment]
@qlzh727  Can you help me with performing the unit test ? I am trying it for a long time, i am facing some problems.

**Comment by qlzh727** [comment]
I would expect some new unit test being added for this new param, and I didn't find that in this PR.

Do u have any trouble for writing new unit test? You can follow the existing example in the same package for the fit() method with class weights.

**Comment by ayushmankumar7** [comment]
I tried to write the unit test but got this error : TypeError: evaluate() got an unexpected keyword argument 'class_weight'

I couldn't get how to solve this.

**Comment by qlzh727** [comment]
I think the current PR looks reasonable. What's the issue u are seeing in the test? Btw, I am kicking off a test on our side, and we can see the test result soon.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Please update the API docstring for the new class_weight parameter.

**Comment by ayushmankumar7** [comment]
@qlzh727  I updated the docstring. Thank you.

**Comment by gbaned** [comment]
@ayushmankumar7 Can you please check build failures. Thanks!

**Review by qlzh727 — CHANGES_REQUESTED** [review]
The unit test is failed since this is an API interface change. You will need to regenerate API pbtxt. Please see error log in https://source.cloud.google.com/results/invocations/421e7975-eac7-47ca-b5f6-a984dbe66e26/targets.

**Comment by mariemabid** [comment]
Is there any solution please to add class_weight to the evaluate() method?

**Comment by ayushmankumar7** [comment]
@qlzh727  I tried updating the API .pbtxt ... #38350 issue.  I am working on it. Will update it asap. Thank you.

**Comment by ayushmankumar7** [comment]
@qlzh727 Please check. I updated the golden files. Successfully regenerated the .pbtxt files. Thank you for the guidance.

File: tensorflow/python/keras/engine/training.py [hunk 1]
```diff
@@ -1434,6 +1441,7 @@ def evaluate_generator(self,
         workers=workers,
         use_multiprocessing=use_multiprocessing,
         verbose=verbose,
+        class_weight=class_weight,
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training.py:1444]: evaluate_generator() is missing the class_weight param, which cause test failure. Please check the test log.

Comment (ayushmankumar7) [reply tensorflow/python/keras/engine/training.py:1445]: I added the class_weight param in evaluate_generator()

**Comment by ayushmankumar7** [comment]
@qlzh727 I made the requested changes

**Review by qlzh727 — CHANGES_REQUESTED** [review]
The test is still failing, please check log at https://source.cloud.google.com/results/invocations/8be8d9be-589a-4d2c-b4a1-f71ecc980469/log

**Comment by ayushmankumar7** [comment]
@qlzh727 I am sorry for that small mistake. I missed a comma. I fixed it. Thank you.

**Comment by gbaned** [comment]
@ayushmankumar7 Can you please fix build failures ? Thanks!

**Comment by ayushmankumar7** [comment]
@gbaned  I am encountering this error : 

Screenshot from 2020-04-22 20-29-32

I am not sure how to solve it. As suggested, i updated the golden file, wrote unit tests. If i get some help, i will be able to make the changes succesfully. Thank you.

**Comment by qlzh727** [comment]
Why we have the last commit for revert back? https://github.com/tensorflow/tensorflow/pull/37425/commits/0baa946a97098e51f1bb662b551ec9f264a5c0b6

**Comment by ayushmankumar7** [comment]
@qlzh727  I went though the master branch, i saw that i added 'class_weight' to evaluate_generator which wasn't the objective of the PR. So, i reverted it back to the what it was initially and concentrating only on evaluate.

**Comment by qlzh727** [comment]
I think the eval and eval_generator should have the same API interface for consistency.

**Comment by ayushmankumar7** [comment]
Okay, then i am making the changes. Thanks

**Comment by ayushmankumar7** [comment]
@qlzh727 I made the requested changes.

File: tensorflow/python/keras/engine/training_test.py [hunk 2]
```diff
@@ -1754,6 +1754,78 @@ def test_sample_weight_tensor(self):
...
+        epochs=epochs // 2,
+        verbose=0,
+        class_weight=class_weight,
+        validation_split=0.1)
+
+    model.train_on_batch(
+        x_train[:batch_size], y_train[:batch_size], class_weight=class_weight)
+    ref_score = model.evaluate(x_test, y_test, class_weight=class_weight, verbose=0)
+    score = model.evaluate(
+        x_test[test_ids, :], y_test[test_ids, :], class_weight=class_weight, verbose=0)
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training_test.py:1825]: I think this is exceeding the line length limits. Please check the sanity test failure log.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Comment by ayushmankumar7** [comment]
@qlzh727 I solved it. Please have a look. Thank you.

**Comment by ayushmankumar7** [comment]
What's the status of this PR ? @gbaned

**Review by qlzh727 — APPROVED** [review]

**Comment by ayushmankumar7** [comment]
@qlzh727  Thanks for the approval :)

**Comment by qlzh727** [comment]
Sorry for the long wait. I checked with Francois about the PR, and seems that it is expected to not having class_weights for model.evaluate.

Quote from Francois:

"The reason we don't support class weights in `evaluate` is that the class_weight argument represents sample weights that are computed from the labels, but the labels should not be an input to the model during evaluation. During training this is fine, but during evaluation this represents a data leak from the labels to your metrics. If you used class weighting in `evaluate`, your score would not be reproducible on real test data (when you don't have the labels).

So this is conceptually wrong."

I am closing this PR since it is working as intended.

**Pull request closed without merging**